libs/logging/src/main/java/org/elasticsearch/logging/Logger.java

```diff
@@ -0,0 +1,732 @@
+    /**
+     * Logs a message with parameters at debug level.
+     *
+     * @param message the message to log
+     * @param p0      parameter to the message.
+     * @param p1      parameter to the message.
+     * @param p2      parameter to the message.
+     * @param p3      parameter to the message.
+     * @param p4      parameter to the message.
+     */
+    void debug(String message, Object p0, Object p1, Object p2, Object p3, Object p4);
+
+    /**
+     * Logs a message with parameters at debug level.
+     *
+     * @param message the message to log
+     * @param p0      parameter to the message.
+     * @param p1      parameter to the message.
+     * @param p2      parameter to the message.
+     * @param p3      parameter to the message.
+     * @param p4      parameter to the message.
+     * @param p5      parameter to the message.
+     */
+    void debug(String message, Object p0, Object p1, Object p2, Object p3, Object p4, Object p5);
+
+    /**
+     * Logs a message with parameters at the {@link org.elasticsearch.logging.Level#DEBUG DEBUG} level.
+     *
+     * @param message the message to log
+     * @param params  parameters to the message.
+     */
+    void debug(String message, Object... params);
+
+    // -- trace
+
+    /**
+     * Logs a lazily supplied String message. If the logger is currently enabled for
+     * {@link org.elasticsearch.logging.Level#FATAL FATAL} level, then a message is logged
+     * that is the result produced by the given supplier function.
+     * @param messageSupplier A function, which when called, produces the desired log String message;
+     */
+    void trace(java.util.function.Supplier<String> messageSupplier);
+
+    /**
+     * Logs a lazily supplied String message associated with a given throwable. If the logger is currently
+     * enabled for {@link org.elasticsearch.logging.Level#FATAL FATAL}  level, then a message is logged
+     * that is the result produced by the given supplier function.
+     * @param messageSupplier A function, which when called, produces the desired log String message;
+     * @param throwable       A Throwable associated with the log message.
+     */
+    void trace(java.util.function.Supplier<String> messageSupplier, Throwable throwable);
+
+    /**
+     * Logs a message at the {@link org.elasticsearch.logging.Level#TRACE TRACE} level.
+     *
+     * @param message the message string to be logged
+     */
+    void trace(String message);
+
+    /**
+     * Logs a message at the {@link org.elasticsearch.logging.Level#TRACE TRACE} level.
+     *
+     * @param message   the message string to be logged
+     * @param throwable A Throwable associated with the log message.
+     */
+    void trace(String message, Throwable throwable);
+
+    /**
+     * Logs a message with parameters at trace level.
+     *
+     * @param message the message to log
+     * @param p0      parameter to the message.
+     */
+    void trace(String message, Object p0);
```

Comment:
I wonder if we should omit the single arg variants of these methods for all levels. The reason is it can be ambiguous with the general method taking an array of args. I'm not sure if there is a performance hit for that, but we are talking about logging so I can't imagine a single implicit array creation is going to make a huge difference.

Reply:
good point, I will remove them

Reply:
Yeah I think we should. We can always add them back.